File: static/solutions/cpp/Max-sum_subarray.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <random>
#include <utility>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
int FindMaximumSubarray(const vector<int>& A) {
  int min_sum = 0, sum = 0, max_sum = 0;
  for (int i = 0; i < A.size(); ++i) {
    sum += A[i];
    if (sum < min_sum) {
      min_sum = sum;
    }
    if (sum - min_sum > max_sum) {
      max_sum = sum - min_sum;
    }
  }
  return max_sum;
}
// @exclude

template <typename T>
vector<T> RandVector(int len) {
  vector<T> ret;
  default_random_engine gen((random_device())());
  while (len--) {
    uniform_int_distribution<int> dis(-1000, 1000);
    ret.push_back(dis(gen));
  }
  return ret;
}

template <typename T>
void CheckMaxSum(const vector<T>& A, int max_sum) {
  for (int i = 0; i < A.size(); ++i) {
    T sum = 0;
    for (int j = i; j < A.size(); ++j) {
      sum += A[j];
      assert(sum <= max_sum);
    }
  }
}

void SmallTest() {
  vector<int> B = {1};
  int max_sum = FindMaximumSubarray(B);
  CheckMaxSum(B, max_sum);
  B = {-5};
  max_sum = FindMaximumSubarray(B);
  CheckMaxSum(B, max_sum);
  B = {0};
  max_sum = FindMaximumSubarray(B);
  CheckMaxSum(B, max_sum);
  B = {0, 0};
  max_sum = FindMaximumSubarray(B);
  CheckMaxSum(B, max_sum);
  B = {0, 0, 0};
  max_sum = FindMaximumSubarray(B);
  CheckMaxSum(B, max_sum);
  B = {0, -5, 0};
  max_sum = FindMaximumSubarray(B);
  CheckMaxSum(B, max_sum);
  B = {-2, -1};
  max_sum = FindMaximumSubarray(B);
  CheckMaxSum(B, max_sum);
}

int main(int argc, char* argv[]) {
  SmallTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    vector<int> A;
    if (argc == 1) {
      uniform_int_distribution<int> dis(1, 10000);
      A = RandVector<int>(dis(gen));
    } else if (argc == 2) {
      int n = atoi(argv[1]);
      A = RandVector<int>(n);
    } else {
      for (int i = 1; i < argc; ++i) {
        A.push_back(atoi(argv[i]));
      }
    }
    int max_sum = FindMaximumSubarray(A);
    CheckMaxSum(A, max_sum);
  }
  return 0;
}

File: static/solutions/java/src/main/java/com/epi/MaxSumSubarray.java
package com.epi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class MaxSumSubarray {
  // @include
  public static int findMaximumSubarray(List<Integer> A) {
    int minSum = 0, sum = 0, maxSum = 0;
    for (int i = 0; i < A.size(); ++i) {
      sum += A.get(i);
      if (sum < minSum) {
        minSum = sum;
      }
      if (sum - minSum > maxSum) {
        maxSum = sum - minSum;
      }
    }
    return maxSum;
  }
  // @exclude

  private static List<Integer> randVector(int len) {
    Random r = new Random();
    List<Integer> ret = new ArrayList<>();
    while (len-- != 0) {
      ret.add(r.nextInt(2001) - 1000);
    }
    return ret;
  }

  private static void checkMaxSum(List<Integer> A, int maxSum) {
    for (int i = 0; i < A.size(); ++i) {
      int sum = 0;
      for (int j = i; j < A.size(); ++j) {
        sum += A.get(j);
        assert(sum <= maxSum);
      }
    }
  }

  private static void simpleTest() {
    List<Integer> B = Arrays.asList(1);
    int maxSum = findMaximumSubarray(B);
    checkMaxSum(B, maxSum);
    B = Arrays.asList(-5);
    maxSum = findMaximumSubarray(B);
    checkMaxSum(B, maxSum);
    B = Arrays.asList(0);
    maxSum = findMaximumSubarray(B);
    checkMaxSum(B, maxSum);
    B = Arrays.asList(0, 0);
    maxSum = findMaximumSubarray(B);
    checkMaxSum(B, maxSum);
    B = Arrays.asList(0, 0, 0);
    maxSum = findMaximumSubarray(B);
    checkMaxSum(B, maxSum);
    B = Arrays.asList(0, -5, 0);
    maxSum = findMaximumSubarray(B);
    checkMaxSum(B, maxSum);
    B = Arrays.asList(-2, -1);
    maxSum = findMaximumSubarray(B);
    checkMaxSum(B, maxSum);
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    for (int times = 0; times < 10000; ++times) {
      List<Integer> A;
      if (args.length == 0) {
        A = randVector(r.nextInt(10000) + 1);
      } else if (args.length == 1) {
        int n = Integer.parseInt(args[0]);
        A = randVector(n);
      } else {
        A = new ArrayList<>();
        for (String arg : args) {
          A.add(Integer.parseInt(arg));
        }
      }
      int maxSum = findMaximumSubarray(A);
      checkMaxSum(A, maxSum);
    }
  }
}
